mne/time_frequency/_stockwell.py

```diff
@@ -55,51 +54,45 @@ def _precompute_st_windows(M, start_f, stop_f, sfreq, width):
                       np.exp(-0.5 * (1. / k ** 2.) * (f ** 2.) * tw ** 2.))
         window /= window.sum()  # normalisation
         windows[i_f] = fftpack.fft(window)
-    return M, f_range, windows
+    return windows
 
 
-def _st(x, M, f_range, windows):
-    """Implementation based on Matlab code by Ali Moukadem"""
-    if x.ndim == 1:
-        x = x[np.newaxis, :]
+def _st(x, start_f, windows):
+    """Implementation based on Ali Moukadem Matlab code (only used in tests)"""
+    n_samp = x.shape[-1]
+    ST = np.empty(x.shape[:-1] + (len(windows), n_samp), dtype=np.complex)
+    # do the work
     Fx = fftpack.fft(x)
-    XF = np.c_[Fx, Fx]
-    ST = np.empty((x.shape[0], len(f_range), M), dtype=np.complex)
-    for i_f, (window, f) in enumerate(zip(windows, f_range)):
-        for i in range(len(x)):
-            ST[i, i_f, :] = fftpack.ifft(XF[i, f:f + M] * window)
-    if ST.shape[0] == 1:
-        ST = ST[0]
+    XF = np.concatenate([Fx, Fx], axis=-1)
+    for i_f, window in enumerate(windows):
+        f = start_f + i_f
+        ST[..., i_f, :] = fftpack.ifft(XF[..., f:f + n_samp] * window)
     return ST
 
 
-def _st_power_itc(x, M, f_range, windows, compute_itc, zero_pad, decim,
-                  final_times):
+def _st_power_itc(x, start_f, compute_itc, zero_pad, decim, W):
     """Aux function"""
-    psd = np.zeros((len(f_range), final_times))
-    if compute_itc is True:
-        itc = np.zeros((len(f_range), final_times), dtype=np.complex)
-    else:
-        itc = None
-    for tfr in _st(x, M, f_range, windows):
-        tfr = tfr[..., :-zero_pad]
-        tfr = tfr[..., ::decim]
-        tfr_abs = np.abs(tfr)
-        psd += tfr_abs ** 2
+    n_samp = x.shape[-1]
+    n_out = (n_samp - zero_pad) // decim
+    psd = np.empty((len(W), n_out))
+    itc = np.empty_like(psd) if compute_itc else None
+    X = fftpack.fft(x)
+    XX = np.concatenate([X, X], axis=-1)
+    for i_f, window in enumerate(W):
+        f = start_f + i_f
+        ST = fftpack.ifft(XX[:, f:f + n_samp] * window)
+        TFR = ST[:, :-zero_pad:decim]
+        TFR_abs = np.abs(TFR)
         if compute_itc:
-            itc += tfr / tfr_abs
-
-    n_signals = x.shape[0]
-    psd /= n_signals
-    if compute_itc:
-        itc = np.abs(itc) / n_signals
+            TFR /= TFR_abs
+            itc[i_f] = np.abs(np.mean(TFR, axis=0))
+        TFR_abs *= TFR_abs
+        psd[i_f] = np.mean(TFR_abs, axis=0)
```

Comment:
is it faster this way?

I suspect this change is putting all the TFRs in memory before averaging and this can be a problem as the memory usage can be huge. Even if it allows to run the FFTs in parallel with multithread I am not sure I want to pay the price in memory usage.